### PR TITLE
Add tests for qemu-tcg backend

### PR DESCRIPTION
**Pull request by ereslibre**
This adds tests for the qemu-tcg backend that check that the emulated machine has the expected architecture. This check also works for the rest of hypervisors.

**However**, there is something fishy when running this test case. Please, check the first comment on this PR for the error.

If you have any idea how to tackle this issue I would be glad to update the PR. Thanks!

### CODE REVIEW

**Comment by ereslibre** [comment]
Although I think this PR should work and pass, I'm getting what I think might be unrelated errors. Any help or pointer would be greatly appreciated.

<details>

<summary>Error details</summary>

```
❯ nix build --verbose -L .#hydraJobs.x86_64-linux.qemu-tcg-startup-shutdown
Using saved setting for 'extra-substituters = https://microvm.cachix.org' from ~/.local/share/nix/trusted-settings.json.
Using saved setting for 'extra-trusted-public-keys = microvm.cachix.org-1:oXnBc6hRE3eX5rSYdRyMYXnfzcCxC7yKPTbZXALsqys=' from ~/.local/share/nix/trusted-settings.json.
warning: ignoring untrusted substituter 'https://microvm.cachix.org', you are not a trusted user.
Run `man nix.conf` for more information on the `substituters` configuration option.
this derivation will be built:
  /nix/store/90llckn2x023wqh2cr3j3gxgjmyisivw-microvm-test-startup-shutdown.drv
building '/nix/store/90llckn2x023wqh2cr3j3gxgjmyisivw-microvm-test-startup-shutdown.drv'...
microvm-test-startup-shutdown> chattr: Operation not supported while setting flags on var.img
microvm-test-startup-shutdown> mke2fs 1.47.0 (5-Feb-2023)
microvm-test-startup-shutdown> Discarding device blocks: done
microvm-test-startup-shutdown> Creating filesystem with 32768 1k blocks and 8192 inodes
microvm-test-startup-shutdown> Filesystem UUID: 64e2c280-e8df-4f7e-a87d-1791c42385a4
microvm-test-startup-shutdown> Superblock backups stored on blocks:
microvm-test-startup-shutdown>  8193, 24577
microvm-test-startup-shutdown> Allocating group tables: done
microvm-test-startup-shutdown> Writing inode tables: done
microvm-test-startup-shutdown> Creating journal (4096 blocks): done
microvm-test-startup-shutdown> Writing superblocks and filesystem accounting information: done
microvm-test-startup-shutdown> <<< NixOS Stage 1 >>>
microvm-test-startup-shutdown> loading module virtio_mmio...
microvm-test-startup-shutdown> loading module virtio_pci...
microvm-test-startup-shutdown> loading module virtio_blk...
microvm-test-startup-shutdown> loading module 9pnet_virtio...
microvm-test-startup-shutdown> loading module 9p...
microvm-test-startup-shutdown> loading module virtiofs...
microvm-test-startup-shutdown> loading module dm_mod...
microvm-test-startup-shutdown> running udev...
microvm-test-startup-shutdown> Starting systemd-udevd version 254.6
microvm-test-startup-shutdown> kbd_mode: KDSKBMODE: Inappropriate ioctl for device
microvm-test-startup-shutdown> %Gstarting device mapper and LVM...
microvm-test-startup-shutdown> mounting rootfs on /...
microvm-test-startup-shutdown> mounting /dev/disk/by-label/nix-store on /nix/store...
microvm-test-startup-shutdown> [    4.947356] /dev/disk/by-label/nix-store: Can't open blockdev
microvm-test-startup-shutdown> checking /dev/vdb...
microvm-test-startup-shutdown> fsck (busybox 1.36.1)
microvm-test-startup-shutdown> [fsck.ext4 (1) -- /mnt-root/var] fsck.ext4 -a /dev/vdb
microvm-test-startup-shutdown> /dev/vdb is mounted.
microvm-test-startup-shutdown> e2fsck: Cannot continue, aborting.
microvm-test-startup-shutdown> fsck on /dev/vdb failed.
microvm-test-startup-shutdown> An error occurred in stage 1 of the boot process, which must mount the
microvm-test-startup-shutdown> root filesystem on `/mnt-root' and then start stage 2.  Press one
microvm-test-startup-shutdown> of the following keys:
microvm-test-startup-shutdown>   r) to reboot immediately
microvm-test-startup-shutdown>   *) to ignore the error and continue
[1/0/1 built] building microvm-test-startup-shutdown
```
</details>

Trying to change the mountpoint from `/var` to `/var/.output` on [`startup-shutdown.nix`](https://github.com/astro/microvm.nix/blob/d5553b1388f2947915c4cec6249b89474046573a/checks/startup-shutdown.nix#L18) gets a little further (!! -- no clue why, diff on top of this PR), but still doesn't work:

<details>

<summary>Error details</summary>

```
❯ nix build --verbose -L .#hydraJobs.x86_64-linux.qemu-tcg-startup-shutdown
warning: Git tree '/home/ereslibre/projects/astro/microvm.nix' is dirty
Using saved setting for 'extra-substituters = https://microvm.cachix.org' from ~/.local/share/nix/trusted-settings.json.
Using saved setting for 'extra-trusted-public-keys = microvm.cachix.org-1:oXnBc6hRE3eX5rSYdRyMYXnfzcCxC7yKPTbZXALsqys=' from ~/.local/share/nix/trusted-settings.json.
warning: ignoring untrusted substituter 'https://microvm.cachix.org', you are not a trusted user.
Run `man nix.conf` for more information on the `substituters` configuration option.
this derivation will be built:
  /nix/store/zka9lz2sdansmf7rcrdz5nd7k46ghmjg-microvm-test-startup-shutdown.drv
building '/nix/store/zka9lz2sdansmf7rcrdz5nd7k46ghmjg-microvm-test-startup-shutdown.drv'...
microvm-test-startup-shutdown> chattr: Operation not supported while setting flags on var.img
microvm-test-startup-shutdown> mke2fs 1.47.0 (5-Feb-2023)
microvm-test-startup-shutdown> Discarding device blocks: done
microvm-test-startup-shutdown> Creating filesystem with 32768 1k blocks and 8192 inodes
microvm-test-startup-shutdown> Filesystem UUID: f6e8c29b-d0fa-4f8c-a58a-f91323b7967c
microvm-test-startup-shutdown> Superblock backups stored on blocks:
microvm-test-startup-shutdown>  8193, 24577
microvm-test-startup-shutdown> Allocating group tables: done
microvm-test-startup-shutdown> Writing inode tables: done
microvm-test-startup-shutdown> Creating journal (4096 blocks): done
microvm-test-startup-shutdown> Writing superblocks and filesystem accounting information: done
microvm-test-startup-shutdown> <<< NixOS Stage 1 >>>
microvm-test-startup-shutdown> loading module virtio_mmio...
microvm-test-startup-shutdown> loading module virtio_pci...
microvm-test-startup-shutdown> loading module virtio_blk...
microvm-test-startup-shutdown> loading module 9pnet_virtio...
microvm-test-startup-shutdown> loading module 9p...
microvm-test-startup-shutdown> loading module virtiofs...
microvm-test-startup-shutdown> loading module dm_mod...
microvm-test-startup-shutdown> running udev...
microvm-test-startup-shutdown> Starting systemd-udevd version 254.6
microvm-test-startup-shutdown> kbd_mode: KDSKBMODE: Inappropriate ioctl for device
microvm-test-startup-shutdown> %Gstarting device mapper and LVM...
microvm-test-startup-shutdown> mounting rootfs on /...
microvm-test-startup-shutdown> mounting /dev/disk/by-label/nix-store on /nix/store...
microvm-test-startup-shutdown> [    4.945547] /dev/disk/by-label/nix-store: Can't open blockdev
microvm-test-startup-shutdown> <<< NixOS Stage 2 >>>
microvm-test-startup-shutdown> running activation script...
microvm-test-startup-shutdown> setting up /etc...
microvm-test-startup-shutdown> error: creating directory '/nix/store/.links': Read-only file system
microvm-test-startup-shutdown> starting systemd...
microvm-test-startup-shutdown> Welcome to NixOS 24.05 (Uakari)!
microvm-test-startup-shutdown> [  OK  ] Created slice Slice /system/getty.
microvm-test-startup-shutdown> [  OK  ] Created slice Slice /system/modprobe.
microvm-test-startup-shutdown> [  OK  ] Created slice Slice /system/serial-getty.
microvm-test-startup-shutdown> [  OK  ] Created slice Slice /system/systemd-fsck.
microvm-test-startup-shutdown> [  OK  ] Created slice User and Session Slice.
microvm-test-startup-shutdown> [  OK  ] Started Dispatch Password Requests to Console Directory Watch.
microvm-test-startup-shutdown> [  OK  ] Started Forward Password Requests to Wall Directory Watch.
microvm-test-startup-shutdown> [  OK  ] Reached target Local Encrypted Volumes.
microvm-test-startup-shutdown> [  OK  ] Reached target Containers.
microvm-test-startup-shutdown> [  OK  ] Reached target Path Units.
microvm-test-startup-shutdown> [  OK  ] Reached target Remote File Systems.
microvm-test-startup-shutdown> [  OK  ] Reached target Slice Units.
microvm-test-startup-shutdown> [  OK  ] Reached target Swaps.
microvm-test-startup-shutdown> [  OK  ] Listening on Process Core Dump Socket.
microvm-test-startup-shutdown> [  OK  ] Listening on Journal Socket (/dev/log).
microvm-test-startup-shutdown> [  OK  ] Listening on Journal Socket.
microvm-test-startup-shutdown> [  OK  ] Listening on Network Service Netlink Socket.
microvm-test-startup-shutdown> [  OK  ] Listening on Userspace Out-Of-Memory (OOM) Killer Socket.
microvm-test-startup-shutdown> [  OK  ] Listening on udev Control Socket.
microvm-test-startup-shutdown> [  OK  ] Listening on udev Kernel Socket.
microvm-test-startup-shutdown>          Mounting Huge Pages File System...
microvm-test-startup-shutdown>          Mounting POSIX Message Queue File System...
microvm-test-startup-shutdown>          Mounting Kernel Debug File System...
microvm-test-startup-shutdown>          Starting Create List of Static Device Nodes...
microvm-test-startup-shutdown>          Starting Load Kernel Module configfs...
microvm-test-startup-shutdown>          Starting Load Kernel Module drm...
microvm-test-startup-shutdown>          Starting Load Kernel Module efi_pstore...
microvm-test-startup-shutdown>          Starting Load Kernel Module fuse...
microvm-test-startup-shutdown>          Starting Create SUID/SGID Wrappers...
microvm-test-startup-shutdown>          Starting Journal Service...
microvm-test-startup-shutdown>          Starting Load Kernel Modules...
microvm-test-startup-shutdown>          Starting Remount Root and Kernel File Systems...
microvm-test-startup-shutdown>          Starting Coldplug All udev Devices...
microvm-test-startup-shutdown> [  OK  ] Mounted Huge Pages File System.
microvm-test-startup-shutdown> [  OK  ] Mounted POSIX Message Queue File System.
microvm-test-startup-shutdown> [  OK  ] Mounted Kernel Debug File System.
microvm-test-startup-shutdown> [  OK  ] Finished Create List of Static Device Nodes.
microvm-test-startup-shutdown> [  OK  ] Finished Load Kernel Module configfs.
microvm-test-startup-shutdown> [  OK  ] Finished Load Kernel Module drm.
microvm-test-startup-shutdown> [  OK  ] Finished Load Kernel Module efi_pstore.
microvm-test-startup-shutdown> [  OK  ] Finished Load Kernel Module fuse.
microvm-test-startup-shutdown> [  OK  ] Finished Remount Root and Kernel File Systems.
microvm-test-startup-shutdown> [  OK  ] Finished Load Kernel Modules.
microvm-test-startup-shutdown>          Mounting FUSE Control File System...
microvm-test-startup-shutdown>          Mounting Kernel Configuration File System...
microvm-test-startup-shutdown>          Starting Firewall...
microvm-test-startup-shutdown>          Starting Load/Save OS Random Seed...
microvm-test-startup-shutdown>          Starting Apply Kernel Variables...
microvm-test-startup-shutdown>          Starting Create Static Device Nodes in /dev gracefully...
microvm-test-startup-shutdown> [  OK  ] Started Journal Service.
microvm-test-startup-shutdown> [  OK  ] Mounted FUSE Control File System.
microvm-test-startup-shutdown> [  OK  ] Mounted Kernel Configuration File System.
microvm-test-startup-shutdown> [  OK  ] Finished Load/Save OS Random Seed.
microvm-test-startup-shutdown> [  OK  ] Finished Apply Kernel Variables.
microvm-test-startup-shutdown>          Starting Flush Journal to Persistent Storage...
microvm-test-startup-shutdown> [  OK  ] Finished Create Static Device Nodes in /dev gracefully.
microvm-test-startup-shutdown>          Starting Create Static Device Nodes in /dev...
microvm-test-startup-shutdown> [  OK  ] Finished Flush Journal to Persistent Storage.
microvm-test-startup-shutdown> [  OK  ] Finished Create Static Device Nodes in /dev.
microvm-test-startup-shutdown> [  OK  ] Reached target Preparation for Local File Systems.
microvm-test-startup-shutdown>          Starting Rule-based Manager for Device Events and Files...
microvm-test-startup-shutdown> [  OK  ] Finished Create SUID/SGID Wrappers.
microvm-test-startup-shutdown> [  OK  ] Started Rule-based Manager for Device Events and Files.
microvm-test-startup-shutdown> [*     ] (1 of 5) A start job is running for…by-label/nix-store (7s / 1min 30s)
microvm-test-startup-shutdown> [**    ] (1 of 5) A start job is running for…by-label/nix-store (8s / 1min 30s)
microvm-test-startup-shutdown> [  OK  ] Finished Firewall.
microvm-test-startup-shutdown> [  OK  ] Reached target Preparation for Network.
microvm-test-startup-shutdown> [  OK  ] Reached target All Network Interfaces (deprecated).
microvm-test-startup-shutdown>          Starting Network Configuration...
microvm-test-startup-shutdown> [  OK  ] Listening on Load/Save RF Kill Switch Status /dev/rfkill Watch.
microvm-test-startup-shutdown> [  OK  ] Started Network Configuration.
microvm-test-startup-shutdown> [***   ] (1 of 4) A start job is running for…y-label/nix-store (12s / 1min 30s)
microvm-test-startup-shutdown> [  OK  ] Finished Coldplug All udev Devices.
microvm-test-startup-shutdown> [  OK  ] Found device /dev/ttyAMA0.
microvm-test-startup-shutdown> [  OK  ] Found device /dev/vdb.
microvm-test-startup-shutdown>          Starting File System Check on /dev/disk/by-label/nix-store...
microvm-test-startup-shutdown>          Starting File System Check on /dev/vdb...
microvm-test-startup-shutdown> [  OK  ] Finished File System Check on /dev/disk/by-label/nix-store.
microvm-test-startup-shutdown> [  OK  ] Finished File System Check on /dev/vdb.
microvm-test-startup-shutdown>          Mounting /var/.output...
microvm-test-startup-shutdown> [FAILED] Failed to mount /var/.output.
microvm-test-startup-shutdown> See 'systemctl status var-.output.mount' for details.
microvm-test-startup-shutdown> [DEPEND] Dependency failed for Local File Systems.
microvm-test-startup-shutdown> [  OK  ] Stopped Dispatch Password Requests to Console Directory Watch.
microvm-test-startup-shutdown> [  OK  ] Stopped Forward Password Requests to Wall Directory Watch.
microvm-test-startup-shutdown> [  OK  ] Reached target Timer Units.
microvm-test-startup-shutdown> [  OK  ] Reached target Login Prompts.
microvm-test-startup-shutdown> [  OK  ] Reached target Socket Units.
microvm-test-startup-shutdown> [  OK  ] Started Emergency Shell.
microvm-test-startup-shutdown> [  OK  ] Reached target Emergency Mode.
microvm-test-startup-shutdown>          Starting Create Volatile Files and Directories...
microvm-test-startup-shutdown> [  OK  ] Finished Create Volatile Files and Directories.
microvm-test-startup-shutdown>          Starting Rebuild Journal Catalog...
microvm-test-startup-shutdown>          Starting Userspace Out-Of-Memory (OOM) Killer...
microvm-test-startup-shutdown>          Starting Network Name Resolution...
microvm-test-startup-shutdown>          Starting Network Time Synchronization...
microvm-test-startup-shutdown>          Starting Record System Boot/Shutdown in UTMP...
microvm-test-startup-shutdown> [  OK  ] Finished Record System Boot/Shutdown in UTMP.
microvm-test-startup-shutdown> [  OK  ] Finished Rebuild Journal Catalog.
microvm-test-startup-shutdown>          Starting Update is Completed...
microvm-test-startup-shutdown> [  OK  ] Finished Update is Completed.
microvm-test-startup-shutdown> [  OK  ] Started Userspace Out-Of-Memory (OOM) Killer.
microvm-test-startup-shutdown> [  OK  ] Started Network Time Synchronization.
microvm-test-startup-shutdown> You are in emergency mode. After logging in, type "journalctl -xb" to view
microvm-test-startup-shutdown> system logs, "systemctl reboot" to reboot, or "exit"
microvm-test-startup-shutdown> to continue bootup.
microvm-test-startup-shutdown> Cannot open access to console, the root account is locked.
microvm-test-startup-shutdown> See sulogin(8) man page for more details.
microvm-test-startup-shutdown> Press Enter to continue.
[1/0/1 built] building microvm-test-startup-shutdown
```
<details>

**Comment by ereslibre** [comment]
I think this might be related to the [`virt` generic virtual platform default behavior of producing a DTB](https://qemu-project.gitlab.io/qemu/system/arm/virt.html#hardware-configuration-information-for-bare-metal-programming), and the kernel loading it, what is different than the regular run with the `microvm` virtual platform, [that does not do any of this](https://www.qemu.org/docs/master/system/i386/microvm.html).

I will continue checking to see if we can have a similar behavior with the `virt` virtual platform.

File: lib/default.nix [hunk 1]
```diff
@@ -30,21 +30,24 @@ rec {
...
+      (nixpkgs-lib.optionalString autoCreate ''
+      PATH=$PATH:${with pkgs.buildPackages; lib.makeBinPath [ coreutils util-linux e2fsprogs ]}
 
       if [ ! -e '${image}' ]; then
         touch '${image}'
         # Mark NOCOW
         chattr +C '${image}' || true
         fallocate -l${toString size}MiB '${image}'
-        mkfs.${fsType} '${image}'
+        mkfs.${fsType} ${pkgs.lib.optionalString (label != null) "-L '${label}'"} '${image}'
```
Comment (ereslibre) [inline lib/default.nix]: Different `fsTypes` might have different arguments for labeling volumes. Should I make this parametric at least for the most common ones? WDYT?

E.g. `mkfs.ext4` has `-L` whereas `mkfs.vfat` has `-n`.

**Comment by ereslibre** [comment]
This is working as expected now. Under the `virt` generic virtual platform, the nix store is for some reason under `/dev/vdb` whereas under the `microvm` generic virtual platform, it is under `/dev/vda`. This makes the identification of storage devices brittle when using their letter identifiers.

For this reason, a new optional "label" attribute is added to volumes, only applicable when `autoCreate` is provided. If this attribute is provided, the volume will be identified by `/dev/disk/by-label` instead of `/dev/vdN`.

File: lib/default.nix (hunk 1 above)
Comment (astro) [reply lib/default.nix]: Yes, please do so. I would like to have at least ext[234], xfs, and vfat.

**Review by astro — CHANGES_REQUESTED** [review]
I considered mounting volumes by label before but it had be optional because sometimes you just can't rely on those labels. Now I see that it doesn't inflate the code too much, and that makes me very happy.

Many thanks for the additional test! Did you see the `ci` branch I am currently working on?

File: lib/default.nix (hunk 1 above)
Comment (ereslibre) [reply lib/default.nix]: Done :)

**Comment by ereslibre** [comment]
> I considered mounting volumes by label before but it had be optional because sometimes you just can't rely on those labels. Now I see that it doesn't inflate the code too much, and that makes me very happy.

Super :)

> Many thanks for the additional test! Did you see the ci branch I am currently working on?

I didn't realize! Let  me know if you prefer to wait for this to merge, or if you want me to do some changes on this PR related to that ongoing work.

Thanks a lot for the review!

**Comment by astro** [comment]
Sorry, this test can't run in a Github free tier CI pipeline because it compiles for too long. I'd like to wait for ideas instead of throwing money at it. :-)

**Comment by ereslibre** [comment]
> Sorry, this test can't run in a Github free tier CI pipeline because it compiles for too long. I'd like to wait for ideas instead of throwing money at it. :-)

I can look into what makes the compilation slow. I have this on my machine as well, given there's something that needs to be cross-compiled and is not in the NixOS cache. I believe this is the main issue.

This is not yet running, right? Where is this being built on CI as of today?

I think it would make sense to have a "slow tests" attribute or something like that, that is only evaluated from time to time, and not on every single PR or commit. I would be fine to take care of those issues and bisect myself to find the culprit of the non-passing test. These tests could also be non-blocking for a new "release" if you will and if you are fine with it.

**Pull request merged**